clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -404,6 +411,12 @@ private void maybeTriggerWakeup() {
             throw new WakeupException();
         }
     }
+    
+    private void maybeThrowInterruptException() {
```

Comment:
It seems fine to me for InterruptException to be a possibility on async callbacks. When the polling thread is interrupted it can't make any progress anyway, due to the way the NIO Selector treats interrupts (returns immediately, doesn't clear interrupted state), so quitting early makes sense I think.

Ignoring interrupts while closing would reintroduce the bug this is trying to solve, since commitSync loops forever if the polling thread is interrupted.

I'm not sure I understand why there is usually a race between calls to `close()` and `wakeup()`? Took a look at the KafkaConsumer Javadoc, and I can't really tell why the example at https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L420 is implemented the way it is. It seems like the `closed` AtomicBoolean is only necessary so we can log WakeupException if it happens unexpectedly. Changing the while loop to be `while (true)` seems like it would avoid the potential race.

```
public void run() {
         try {
             consumer.subscribe(Arrays.asList("topic"));
             while (true) {
                 ConsumerRecords records = consumer.poll(10000);
                 // Handle new records
             }
         } catch (WakeupException | InterruptException e) {
             // Ignore exception if closing
             if (!closed.get()) throw e;
         } finally {
             consumer.close();
         }
     }

     // Shutdown hook which can be called from a separate thread
     public void shutdown() {
         closed.set(true); // or maybe just remove this field entirely?
         consumer.wakeup(); // or interrupt, either is fine
     }
```

I'll add catches to the places where WakeupException is caught.